crates/polars-core/src/series/ops/reshape.rs

```diff
@@ -89,31 +89,58 @@ impl Series {
+        if dimensions.is_empty() {
+            polars_bail!(InvalidOperation: "at least one dimension must be specified")
+        }
+
         let mut dims = dimensions.iter().copied().collect::<VecDeque<_>>();
 
         let leaf_array = self.get_leaf_array();
-        let size = leaf_array.len() as i64;
-
-        // Infer dimension
-        if dims.contains(&-1) {
-            let infer_dims = dims.iter().filter(|d| **d == -1).count();
-            polars_ensure!(infer_dims == 1, InvalidOperation: "can only infer single dimension, found {}", infer_dims);
-
-            let mut prod = 1;
-            for &dim in &dims {
-                if dim != -1 {
-                    prod *= dim;
+        let size = leaf_array.len();
+
+        let mut total_dim_size = 1;
+        let mut infer_dim_index: Option<usize> = None;
+        for (index, &dim) in dims.iter().enumerate() {
+            if dim > 0 {
+                total_dim_size *= dim as usize;
+            } else if dim == 0 {
+                if index != 0 {
+                    polars_bail!(
+                        InvalidOperation: "cannot reshape array into shape containing a zero dimension after the first: {}",
+                        format_tuple!(dims)
+                    );
                 }
+                total_dim_size = 0;
+                // We can early exit here, as empty arrays will error with multiple dimensions,
+                // and non-empty arrays will error when the first dimension is zero.
+                break;
+            } else if infer_dim_index.is_none() {
+                infer_dim_index = Some(index);
+            } else {
+                polars_bail!(InvalidOperation: "can only specify one unknown dimension");
             }
-            polars_ensure!(size % prod == 0, InvalidOperation: "cannot reshape array of size {} into shape: {}", size, format_tuple!(dims));
-            let inferred_value = size / prod;
-            for dim in &mut dims {
-                if *dim == -1 {
-                    *dim = inferred_value;
-                    break;
-                }
+        }
+
+        if size == 0 {
+            if dims.len() > 1 || (infer_dim_index.is_none() && total_dim_size != 0) {
+                polars_bail!(InvalidOperation: "cannot reshape empty array into shape {}", format_tuple!(dims))
             }
+        } else if total_dim_size == 0 {
+            polars_bail!(InvalidOperation: "cannot reshape non-empty array into shape containing a zero dimension: {}", format_tuple!(dims))
+        } else {
+            polars_ensure!(
+                size % total_dim_size == 0,
+                InvalidOperation: "cannot reshape array of size {} into shape {}", size, format_tuple!(dims)
+            );
+        }
+
+        // Infer dimension
+        if let Some(index) = infer_dim_index {
+            let inferred_dim = if size == 0 { 0 } else { size / total_dim_size };
```

Comment:
Shouldn't we raise here if `size==0`? I think we cannot infer the dimension in that case.

Reply:
In case there is an empty Series and we reshape with `-1`, we can infer that the dimension should be 0.

Actually I don't think we need this branch anymore because `total_dim_size` will be one. Updating 👍